egs/wsj/s5/steps/nnet3/components.py

```diff
@@ -213,6 +238,24 @@ def AddSigmoidLayer(config_lines, name, input, self_repair_scale = None):
     return {'descriptor':  '{0}_sigmoid'.format(name),
             'dimension': input['dimension']}
 
+def AddUnfoldedRnnLayer(config_lines, name, input, output_dim, ng_affine_options = "", self_repair_scale = None):
+    components = config_lines['components']
+    component_nodes = config_lines['component-nodes']
+
+    components.append("component name={0}_affine type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input[0]['dimension'] + output_dim, output_dim, ng_affine_options))
+    self_repair_string = "self-repair-scale={0:.10f}".format(self_repair_scale) if self_repair_scale is not None else ''
+    components.append("component name={0}_tanh type=TanhComponent dim={1}".format(name, output_dim, self_repair_string))
+    # create a dummy node as a recurrent input to the first frame (it is expected to be 0 because it would not be computable in the computation graph) 
+    component_nodes.append("component-node name={0}_tanh_dummy component={0}_tanh input=IfDefined(ReplaceIndex({0}_tanh_t, x, -1))".format(name))
```

Comment:
It shouldn't be necessary to add a dummy node-- just take the input from a
node that does exist, and has the right dimension, but give it an invalid
index x=-1.


On Tue, Jun 14, 2016 at 9:04 PM, Yiming Wang <notifications@github.com>
wrote:

> In egs/wsj/s5/steps/nnet3/components.py
> <https://github.com/kaldi-asr/kaldi/pull/848#discussion_r67084699>:
>
> > @@ -213,6 +238,24 @@ def AddSigmoidLayer(config_lines, name, input, self_repair_scale = None):
> >      return {'descriptor':  '{0}_sigmoid'.format(name),
> >              'dimension': input['dimension']}
> >
> > +def AddUnfoldedRnnLayer(config_lines, name, input, output_dim, ng_affine_options = "", self_repair_scale = None):
> > +    components = config_lines['components']
> > +    component_nodes = config_lines['component-nodes']
> > +
> > +    components.append("component name={0}_affine type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input[0]['dimension'] + output_dim, output_dim, ng_affine_options))
> > +    self_repair_string = "self-repair-scale={0:.10f}".format(self_repair_scale) if self_repair_scale is not None else ''
> > +    components.append("component name={0}_tanh type=TanhComponent dim={1}".format(name, output_dim, self_repair_string))
> > +    # create a dummy node as a recurrent input to the first frame (it is expected to be 0 because it would not be computable in the computation graph)
> > +    component_nodes.append("component-node name={0}_tanh_dummy component={0}_tanh input=IfDefined(ReplaceIndex({0}_tanh_t, x, -1))".format(name))
>
> It is to made the "dummy" node NOT computable so that it would be a zero
> input to the first frame (just like how "IfDefined" descriptor is used in
> the RNN setting). Will add to the comments.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/848/files/4865d8bc204f63a79c07d489c59e0c325262a994#r67084699>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADJVu2WeIGvh045aWFlkbm_dNr5PD06hks5qL0-2gaJpZM4I1l8l>
> .
>